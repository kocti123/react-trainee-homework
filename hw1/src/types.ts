import { AnyAction } from "redux";
import { ThunkAction } from "redux-thunk";

export interface ITask {
  readonly id: number;
  readonly body: string;
  readonly completed: boolean;
  readonly favourite: boolean;
  readonly isEdit: boolean;
  readonly newBody: string;
  readonly creationDate: Date;
}

export type filters = "all" | "completed" | "favourite" | "notCompleted";

export type TasksState = {
  readonly isLoaded: boolean;
  readonly lastId: number;
  readonly tasks: ITask[];
  readonly currentFilter: filters;
  readonly filtredTasks: ITask[];
};

export type State = {
  readonly tasks: TasksState;
};

export type AppAction = ThunkAction<void, State, unknown, AnyAction>;
