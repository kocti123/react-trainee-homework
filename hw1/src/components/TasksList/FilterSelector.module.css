.filterSelector {
  display: flex;
  border-bottom: 1px solid black;
}

.filterSelector button {
  border: none;
  width: 100%;
  padding-top: 10px;
  padding-bottom: 10px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  cursor: pointer;
}

.filterSelector button:hover {
  background-color: grey;
}

.filterSelector button.selected {
  background-color: black;
  color: white;
}
