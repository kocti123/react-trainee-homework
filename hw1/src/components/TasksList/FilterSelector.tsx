import { useAppDispatch, useAppSelector } from "../../redux";
import { changeFilter } from "../../redux/actions/tasksActions";

import styles from "./FilterSelector.module.css";

function FilterSelector() {
  const dispatch = useAppDispatch();
  const { currentFilter } = useAppSelector((store) => store.tasks);

  const filterFavourite = () => {
    if ("favourite" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("favourite"));
  };
  const filterCompleted = () => {
    if ("completed" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("completed"));
  };
  const filterNotCompleted = () => {
    if ("notCompleted" === currentFilter) {
      dispatch(changeFilter("all"));
      return;
    }
    dispatch(changeFilter("notCompleted"));
  };

  return (
    <div className={styles.filterSelector}>
      <button
        className={currentFilter === "completed" ? styles.selected : undefined}
        onClick={filterCompleted}
      >
        Выполненные задачи
      </button>
      <button
        className={
          currentFilter === "notCompleted" ? styles.selected : undefined
        }
        onClick={filterNotCompleted}
      >
        Задачи в работе
      </button>
      <button
        className={currentFilter === "favourite" ? styles.selected : undefined}
        onClick={filterFavourite}
      >
        Избранные задачи
      </button>
    </div>
  );
}

export default FilterSelector;
