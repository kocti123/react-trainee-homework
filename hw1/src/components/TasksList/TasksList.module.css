.tasksMenu {
  margin-top: 20px;
  width: 400px;
  overflow: hidden;
  border-radius: 20px;
  border: 1px solid black;
}

.tasksList li:last-child {
  margin-bottom: 0px;
  border: none;
}

.tasksList {
  padding: 0;
  margin: 0;
}

.noTasks {
  text-align: center;
}

.loading {
  padding: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.loading p {
  margin-top: 5px;
}
