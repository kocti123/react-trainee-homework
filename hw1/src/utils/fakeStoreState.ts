import { State } from "../types";

export function getFakeStoreState(): State {
  return {
    tasks: {
      isLoaded: true,
      lastId: 1,
      currentFilter: "all",
      tasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
      filtredTasks: [
        {
          id: 1,
          body: "test",
          completed: false,
          favourite: false,
          isEdit: false,
          newBody: "",
          creationDate: new Date(1635696113017),
        },
      ],
    },
  };
}
