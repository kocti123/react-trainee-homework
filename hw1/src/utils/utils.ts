import { AnyAction } from "redux";
import { tasksActionTypes } from "../constants/actionTypes";
import { filters, ITask } from "../types";

export const filterTasks = (tasks: ITask[], filter: filters): ITask[] => {
  switch (filter) {
    case "all":
      return tasks;
    case "favourite":
      return filterFavouriteAndNotCompleted(tasks);
    case "completed":
      return filterCompleted(tasks);
    case "notCompleted":
      return filterNotCompleted(tasks);
    default:
      throw new Error("wrong filter type");
  }
};

export const filterFavouriteAndNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.favourite && !task.completed);
};

export const filterCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => task.completed);
};

export const filterNotCompleted = (tasks: ITask[]): ITask[] => {
  return tasks.filter((task) => !task.completed);
};

export const filterOutById = (tasks: ITask[], id: number): ITask[] => {
  return tasks.filter((task) => task.id !== id);
};

export const findById = (tasks: ITask[], id: number): ITask => {
  return tasks.filter((task) => task.id === id)[0];
};

export const lastIdOfTasks = (tasks: ITask[]): number => {
  return Math.max(...tasks.map((task) => task.id));
};

export const filterAndSortTasks = (
  tasks: ITask[],
  filter: filters
): ITask[] => {
  const filteredTasks = filterTasks(tasks, filter);
  return filteredTasks.sort((a, b) => b.id - a.id);
};

export const getUpdateAction = (id: number, task: ITask): AnyAction => {
  return {
    type: tasksActionTypes.UPDATE_TASK,
    payload: {
      id,
      task: {
        ...task,
      },
    },
  };
};
