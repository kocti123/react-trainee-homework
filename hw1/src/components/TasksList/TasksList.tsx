import { useEffect, useState } from "react";
import { useAppDispatch, useAppSelector } from "../../redux";
import {
  fetchAndSetAllTasks,
  removeTask,
} from "../../redux/actions/tasksActions";
import FilterSelector from "./FilterSelector";
import TaskElement from "../TaskElement/TaskElement";
import styles from "./TasksList.module.css";
import Modal from "../UI/Modal";
import { ITask } from "../../types";
import Spinner from "../UI/Spinner";

function TasksList() {
  const dispatch = useAppDispatch();
  const { filtredTasks, isLoaded } = useAppSelector((store) => store.tasks);
  const [toDelete, setToDelete] = useState<ITask | null>(null);
  const [showToDelete, setShowToDelete] = useState(false);

  useEffect(() => {
    dispatch(fetchAndSetAllTasks());
  }, [dispatch]);

  function deleteHandler(task: ITask): void {
    setToDelete({ ...task });
    setShowToDelete(true);
  }

  function cancelDelete() {
    setShowToDelete(false);
  }

  function deleteTask() {
    if (toDelete) {
      dispatch(removeTask(toDelete.id));
    }
    setShowToDelete(false);
  }

  return (
    <div className={styles.tasksMenu}>
      <FilterSelector />
      {isLoaded ? (
        <>
          {filtredTasks.length !== 0 ? (
            <ul className={styles.tasksList}>
              {filtredTasks.map((task) => {
                return (
                  <TaskElement
                    id={task.id}
                    key={task.id}
                    onDelete={deleteHandler}
                    task={task}
                  />
                );
              })}
            </ul>
          ) : (
            <p className={styles.noTasks}>Нет задач</p>
          )}
        </>
      ) : (
        <div className={styles.loading}>
          <Spinner />
          <p>Загрузка</p>
        </div>
      )}
      <Modal show={showToDelete} title="Удаления задачи" onClose={cancelDelete}>
        <p>Вы действительно хотите удалить задачу?</p>
        <p>Текст задачи:</p>
        <p>{toDelete && toDelete.body}</p>
        <p>Дата создания:</p>
        <p>{toDelete && toDelete.creationDate.toLocaleString()}</p>
        <button onClick={cancelDelete}>Отмена</button>
        <button onClick={deleteTask}>Удалить</button>
      </Modal>
    </div>
  );
}

export default TasksList;
