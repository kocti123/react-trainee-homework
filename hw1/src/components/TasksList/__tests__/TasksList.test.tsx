import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import createMockStore, { MockStoreEnhanced } from "redux-mock-store";
import thunk from "redux-thunk";
import { AppAction, ITask, State } from "../../../types";
import { filterTasks, lastIdOfTasks } from "../../../utils/utils";
import TasksList from "../TasksList";
import styles from "../TasksList.module.css";

const mockActionType = {
  setTasks: "setTasks",
  removeTask: "removeTask",
};

jest.mock("../../../redux/actions/tasksActions", () => ({
  fetchAndSetAllTasks: (): AppAction => (dispatch) => {
    dispatch({
      type: mockActionType.setTasks,
    });
  },
  removeTask:
    (id: number): AppAction =>
      (dispatch) => {
        dispatch({
          type: mockActionType.removeTask,
          payload: id,
        });
      },
}));

const createStore = createMockStore([thunk]);

const tasks: ITask[] = [
  {
    id: 1,
    body: "test",
    completed: true,
    favourite: false,
    isEdit: false,
    newBody: "",
    creationDate: new Date(),
  },
  {
    id: 2,
    body: "test1",
    completed: false,
    favourite: false,
    isEdit: false,
    newBody: "",
    creationDate: new Date(),
  },
];
const state: State = {
  tasks: {
    currentFilter: "completed",
    isLoaded: true,
    tasks: tasks,
    filtredTasks: filterTasks(tasks, "completed"),
    lastId: lastIdOfTasks(tasks),
  },
};

const mountElement = (store: MockStoreEnhanced<unknown, {}>) => {
  return mount(<TasksList />, {
    wrappingComponent: Provider,
    wrappingComponentProps: { store },
  });
};

describe("FilterSelector", () => {
  let modalRoot: HTMLDivElement;

  beforeAll(() => {
    modalRoot = document.createElement("div");
    modalRoot.id = "modal";
    document.body.append(modalRoot);

    window.scrollTo = jest.fn();
  });

  afterAll(() => {
    modalRoot.remove();
    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should match snapshow", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it("should request tasks from server", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    expect(store.getActions()).toEqual([{ type: mockActionType.setTasks }]);

    wrapper.unmount();
  });

  it("should show loading when tasks not loaded", () => {
    const store = createStore({ tasks: { ...state.tasks, isLoaded: false } });
    const wrapper = mountElement(store);

    expect(wrapper.find(`.${styles.loading}`)).toHaveLength(1);

    wrapper.unmount();
  });

  it("should render filtered tasks", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    expect(wrapper.find("TaskElement")).toHaveLength(1);
    expect(wrapper.find("TaskElement").prop<ITask>("task").body).toBe("test");

    wrapper.unmount();
  });

  it("should show message when no tasks", () => {
    const store = createStore({ tasks: { ...state.tasks, filtredTasks: [] } });
    const wrapper = mountElement(store);

    expect(wrapper.find(`.${styles.noTasks}`).text()).toBe("Нет задач");

    wrapper.unmount();
  });

  it("should show modal when trigger deletion", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    act(() => {
      wrapper.find("TaskElement").prop<(task: ITask) => {}>("onDelete")(
        state.tasks.filtredTasks[0]
      );
    });

    wrapper.update();

    expect(wrapper.find("Modal").prop("show")).toBeTruthy();

    wrapper.unmount();
  });

  it("should close modal when Modal triger close", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    act(() => {
      wrapper.find("TaskElement").prop<(task: ITask) => {}>("onDelete")(
        state.tasks.filtredTasks[0]
      );
    });
    wrapper.update();
    expect(wrapper.find("Modal").prop("show")).toBeTruthy();

    act(() => {
      wrapper.find("Modal").prop<() => {}>("onClose")();
    });
    wrapper.update();
    expect(wrapper.find("Modal").prop("show")).toBeFalsy();

    wrapper.unmount();
  });

  it("should close modal when click on cancle button", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    act(() => {
      wrapper.find("TaskElement").prop<(task: ITask) => {}>("onDelete")(
        state.tasks.filtredTasks[0]
      );
    });
    wrapper.update();
    expect(wrapper.find("Modal").prop("show")).toBeTruthy();

    wrapper.find(".cancel").simulate("click");
    expect(wrapper.find("Modal").prop("show")).toBeFalsy();

    wrapper.unmount();
  });

  it("should delete task when click on delete button", () => {
    const store = createStore(state);
    const wrapper = mountElement(store);

    act(() => {
      wrapper.find("TaskElement").prop<(task: ITask) => {}>("onDelete")(
        state.tasks.filtredTasks[0]
      );
    });
    wrapper.update();
    expect(wrapper.find("Modal").prop("show")).toBeTruthy();

    wrapper.find(".delete").simulate("click");
    expect(store.getActions()[1]).toEqual({
      type: mockActionType.removeTask,
      payload: 1,
    });

    wrapper.unmount();
  });
});
