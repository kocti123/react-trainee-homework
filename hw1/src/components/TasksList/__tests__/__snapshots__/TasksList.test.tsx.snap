// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelector should match snapshow 1`] = `ReactWrapper {}`;
